### ISSUE TITLE
add network availability event feed into WebSocket to inform retry loops

### ISSUE THREAD

**Issue by heckj** [automerge/automerge-repo-swift#87]
leverage https://developer.apple.com/documentation/network/nwpathmonitor

**Comment by heckj** [comment]
recent article on NilCoalescing shows a method of exposing this kind of thing into SwiftUI with an AsyncStream, which I thought was interesting, and potentially relevant/useful: https://nilcoalescing.com/blog/ReactToNetworkStatusUpdatesInSwiftUI/

**Comment by jessegrosjean** [comment]
I'm been looking at this, and then realized I didn't know how it should work :)

I guess the key line is:

```
try await Task.sleep(for: .seconds(waitBeforeReconnect))
```

And if Backoff has been going for a while and is a high number... we want to reset (try again right away) if network path becomes satisfied? Is that about what you had imagined?

**Comment by heckj** [comment]
I didn't have a specific heuristic in mind, just that it would "make sense to not try and reconnect" if the NWPath wasn't valid. So the only idea in mind was "don't even try" if the NWPath isn't viable. That may warrant a completely different backoff strategy - something that counts in larger values, but I haven't measured (nor have any intuitive sense) of what kinds of delays there are in NWPath availability state changes, particularly the "OK" -> "NO GOOD" -> "OK" sequence. So my naïve take would be to use the "NO GOOD" state from NWPath as a check before attempting, only attempting IF the path was good, and otherwise just moving on to the next iteration and sticking with the current exponential-ish backoff strategy.

Semi-related, the back-off is (currently) always increasing on each attempt - which if we have a stuttering connection may not be what we want. It's quite possible that we should reset the reconnect counter on a successful reconnection & handshake, so that retries are as fast as possible in the later stages of a stuttering connection.

(the general strategy is all embedded in the logic in [Backoff.swift](https://github.com/automerge/automerge-repo-swift/blob/main/Sources/AutomergeRepo/Networking/Backoff.swift, which returns a number from 0 to 610, max at 610, that's increasing using a fibonacci sequence rather than a straight up exponential backoff. I just liked the curve better. But it maxes out and doesn't keep increasing, so that if you've been disconnected for a long time, you don't get TOO ludicrous a wait for the circuit breaker to reconnect. Right now I'm counting those as seconds, so at max backoff it'll try every ~5 minutes of so)

**Comment by jessegrosjean** [comment]
> I didn't have a specific heuristic in mind, just that it would "make sense to not try and reconnect" if the NWPath wasn't valid. 

I've actually done something a bit different here. For me the confusing part of this (when just testing MeetingNotes) was that if I turned off wifi connect and then turned it back on later... it would sometimes take a very long time to start working again (connect).

So instead this set of changes waits on both the timeout and watch for network connection transitioning from not available to available. And  if it sees that transition then it short circuits the current timeout. The result is socket connection should reconnect immediately after network becomes available

If you think it's important to optimize the other way too (don't try to reconnect if there is no network) I can add that. For this first attempt I figured I would skip that part, we already have backoff and this first change (quicker reconnect) seemed more important to me.

> we should reset the reconnect counter on a successful reconnection & handshake

I think this set of changes does that.
